lib/oidcstrategy.js

```diff
@@ -752,4 +574,468 @@ Strategy.prototype._shouldLoadUserProfile = function shouldLoadUserProfile(issue
+                if (getUserInfoError) {
+                  return next(new InternalOAuthError('failed to fetch user profile', getUserInfoError));
+                }
+
+                log.info('PROFILE LOADED FROM MS IDENTITY', body);
+
+                var userinfoReceived = undefined;
+                // use try / catch around JSON.parse --> could fail unexpectedly
+                try {
+                  userinfoReceived = JSON.parse(body);
+                } catch (ex) {
+                  return next(ex);
+                }
+
+                // make sure the 'sub' in userinfo is the same as the one in 'id_token'
+                if (userinfoReceived.sub !== jwtClaims.sub) {
+                  log.error('sub in userinfo is ' + userinfoReceived.sub + ', but does not match sub in id_token, which is ' + id_token.sub);
+                  return self.fail('sub received in userinfo and id_token do not match');
+                }
+
+                return onProfileLoaded(self, {
+                  req,
+                  sub,
+                  iss,
+                  profile: makeProfileObject(userinfoReceived, body),
+                  jwtClaims,
+                  access_token,
+                  refresh_token,
+                  params,
+                });
+              });
+            }
+
+            // lets do an id_token fallback. We use id_token over userInfo endpoint for now
+            return onProfileLoaded(self, {
+              req,
+              sub,
+              iss,
+              profile: makeProfileObject(jwtClaims, jwtClaimsStr),
+              jwtClaims,
+              access_token,
+              refresh_token,
+              params,
+            });
+          });
+        });
+      });
+  });
+};
+
+/**
+ * prepare the initial authorization request
+ *
+ * @params {Object} req
+ * @params {Function} next
+ */
+Strategy.prototype._flowInitializationHandler = function flowInitializationHandler(req, next) {
+  // The request being authenticated is initiating OpenID Connect
+  // authentication. Prior to redirecting to the provider, configuration will
+  // be loaded. The configuration is typically either pre-configured or
+  // discovered dynamically. When using dynamic discovery, a user supplies
+  // their identifer as input.
+
+  const self = this;
+
+  log.info(`we are in the OpenID Connect Inital Flow`);
+
+  let identifier;
+  if (req.body && req.body[this._identifierField]) {
+    identifier = req.body[this._identifierField];
+  } else if (req.query && req.query[this._identifierField]) {
+    identifier = req.query[this._identifierField];
+  }
+
+  return self.configure(identifier, (configureError, config) => {
+    if (configureError) {
+      return next(configureError);
+    }
+
+    var options = self._options;
+
+    let callbackURL = options.callbackURL || config.callbackURL;
+    if (callbackURL) {
+      const parsed = url.parse(callbackURL);
+      if (!parsed.protocol) {
```

Comment:
@addev-ashish callbackURL is checked at the very beginning when strategy is created